### PR TITLE
Fix `Manager::userHasPermission()`, various fixes and improvements

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️/❌
| New feature?  | ✔️/❌
| Breaks BC?    | ✔️/❌

Additionally:

- Improved Psalm types.
- Raised code coverage to 100%.
- Raised min MSI and fixed mutants.

### CODE REVIEW

File: src/Manager.php [hunk 1]
```diff
@@ -51,24 +51,62 @@ public function userHasPermission(
...
+
+            $userItems = [];
+            foreach ($userItemNames as $itemName) {
+                $userItems[$itemName] = $items[$itemName];
+            }
+
+            foreach ($userItems as $userItem) {
+                foreach ($this->itemsStorage->getAllChildren($userItem->getName()) as $childItem) {
+                    if (array_key_exists($childItem->getName(), $items)) {
+                        $userItems[$childItem->getName()] = $childItem;
```
Comment (vjik) [inline src/Manager.php:71]: If check rules here, we prevent unnecessary calls `getAllChildren`.

File: src/Manager.php [hunk 2]
```diff
@@ -51,24 +51,62 @@ public function userHasPermission(
...
+            $userId = (string) $userId;
+            $userItemNames = $this->assignmentsStorage->filterUserItemNames($userId, array_keys($items));
+
+            $userItems = [];
+            foreach ($userItemNames as $itemName) {
+                $userItems[$itemName] = $items[$itemName];
+            }
+
+            foreach ($userItems as $userItem) {
+                foreach ($this->itemsStorage->getAllChildren($userItem->getName()) as $childItem) {
```
Comment (vjik) [inline src/Manager.php:69]: May be filter childrens by `$items` in items storage implementation allow to optimize this process?

File: src/ManagerInterface.php [hunk 3]
```diff
@@ -113,13 +115,23 @@ public function revoke(string $itemName, int|Stringable|string $userId): self;
      */
     public function revokeAll(int|Stringable|string $userId): self;
 
+    /**
+     * Returns the ite,s that are assigned to the user via {@see assign()}.
```
Comment (vjik) [inline src/ManagerInterface.php:119]:
```suggestion
     * Returns the items that are assigned to the user via {@see assign()}.
```

File: src/ManagerInterface.php [hunk 4]
```diff
@@ -113,13 +115,23 @@ public function revoke(string $itemName, int|Stringable|string $userId): self;
...
+     *
+     * @param int|string|Stringable $userId The user ID.
+     *
+     * @return array All items directly assigned to the user. The array is indexed by the item names.
+     * @psalm-return ItemsIndexedByName
+     */
+    public function getItemsByUserId(int|Stringable|string $userId): array;
```
Comment (vjik) [inline src/ManagerInterface.php:126]: Why need this method?

File: src/Manager.php [hunk 5]
```diff
@@ -281,19 +338,15 @@ public function getDefaultRoleNames(): array
...
-            throw new DefaultRolesNotFoundException("The following default roles were not found: $missingRolesStr.");
-        }
-
-        return $roles;
+        return $this->filterStoredRoles($this->defaultRoleNames);
     }
 
     public function setGuestRoleName(?string $name): self
     {
+        if ($name !== null && !$this->itemsStorage->roleExists($name)) {
```
Comment (vjik) [inline src/Manager.php:346]: It add additional request to storage on each manage configuration. I think better remove this check... If user set invalid guest role name, than manager also throw exception, but in another place.

File: src/ManagerInterface.php (hunk 4 above)
Comment (arogachev) [reply src/ManagerInterface.php:126]: To get all items (both roles and permissions) assigned to user in one call. Before it was possible to get either roles or permissions only.

File: src/Manager.php (hunk 1 above)
Comment (arogachev) [reply src/Manager.php]: https://github.com/yiisoft/rbac/pull/203/commits/6c6c1ee7aeb4aaaf023983d042055cb57d04e339

File: src/Manager.php (hunk 2 above)
Comment (arogachev) [reply src/Manager.php]: https://github.com/yiisoft/rbac/pull/203/commits/6c6c1ee7aeb4aaaf023983d042055cb57d04e339

File: src/Manager.php (hunk 5 above)
Comment (arogachev) [reply src/Manager.php]: https://github.com/yiisoft/rbac/pull/203/commits/4e9bd6fd6519e4bf455c786eb4e27b4a5673267d

**Pull request merged**